src/evict/evict_lru.c

```diff
@@ -1374,29 +1385,31 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 		/*
 		 * It's possible (but unlikely) to visit a page without a read
 		 * generation, if we race with the read instantiating the page.
-		 * Ignore those pages, but set the page's read generation here
-		 * to ensure a bug doesn't somehow leave a page without a read
-		 * generation.
+		 * Set the page's read generation here to ensure a bug doesn't
+		 * somehow leave a page without a read generation.
 		 */
-		if (page->read_gen == WT_READGEN_NOTSET) {
+		if (page->read_gen == WT_READGEN_NOTSET)
 			__wt_cache_read_gen_new(session, page);
-			continue;
-		}
 
 		/* Pages we no longer need (clean or dirty), are found money. */
 		if (page->read_gen == WT_READGEN_OLDEST) {
 			WT_STAT_FAST_CONN_INCR(
 			    session, cache_eviction_pages_queued_oldest);
 			goto fast;
 		}
+
 		if (__wt_page_is_empty(page) ||
 		    F_ISSET(session->dhandle, WT_DHANDLE_DEAD) ||
 		    FLD_ISSET(cache->state, WT_EVICT_STATE_AGGRESSIVE))
 			goto fast;
 
 		/* Skip clean pages if appropriate. */
 		if (!modified && (F_ISSET(conn, WT_CONN_IN_MEMORY) ||
-		    FLD_ISSET(cache->state, WT_EVICT_STATE_DIRTY)))
+		    !FLD_ISSET(cache->state, WT_EVICT_STATE_CLEAN)))
+			continue;
+
+		/* Skip dirty pages if appropriate. */
+		if (modified && !FLD_ISSET(cache->state, WT_EVICT_STATE_DIRTY))
 			continue;
```

Comment:
These changes deserve some thought -- we're now ignoring dirty pages if we're under the dirty target.